src/back/spv/writer.rs

```diff
@@ -1320,6 +1355,234 @@ impl Writer {
+            ty: Handle<crate::Type>,
+            ir_module: &crate::Module,
+        ) -> Result<u32, Error> {
+            match ir_module.types[ty].inner {
+                crate::TypeInner::Struct { ref members, .. } => Ok(members.len() as u32 - 1),
+                crate::TypeInner::Pointer { base, .. } => match ir_module.types[base].inner {
+                    crate::TypeInner::Struct { ref members, .. } => Ok(members.len() as u32 - 1),
+                    _ => Err(Error::Validation("array length expression")),
+                },
+                _ => Err(Error::Validation("array length expression")),
+            }
+        }
+
+        let (structure_id, member_idx) = match ir_function.expressions[array] {
+            crate::Expression::AccessIndex { base, .. } => match ir_function.expressions[base] {
+                crate::Expression::GlobalVariable(handle) => {
+                    let global = &ir_module.global_variables[handle];
+                    let last_idx = last_member_idx(global.ty, ir_module)?;
+                    (self.global_variables[handle.index()].id, last_idx)
+                }
+                crate::Expression::FunctionArgument(index) => {
+                    let argument = &ir_function.arguments[index as usize];
+                    let last_idx = last_member_idx(argument.ty, ir_module)?;
+                    let parameter_id = function.parameter_id(index);
+                    (parameter_id, last_idx)
+                }
+                _ => return Err(Error::Validation("array length expression")),
+            },
+            _ => return Err(Error::Validation("array length expression")),
+        };
+
+        let length_type = LocalType::Value {
+            vector_size: None,
+            kind: crate::ScalarKind::Uint,
+            width: 4,
+            pointer_class: None,
+        };
+        let length_type_id = self.get_type_id(length_type.into())?;
+        let length_id = self.id_gen.next();
+        block.body.push(Instruction::array_length(
+            length_type_id,
+            length_id,
+            structure_id,
+            member_idx,
+        ));
+
+        Ok(length_id)
+    }
+
+    /// Emit code to compute the maximum valid index of a run-time array.
+    ///
+    /// Given `array`, an expression referring to the final member of a struct,
+    /// where the member in question is a runtime-sized array, return the id of
+    /// a SPIR-V insn representing the maximum valid index into that array (that
+    /// is, one less than its length).
+    fn write_runtime_array_limit(
+        &mut self,
+        array: Handle<crate::Expression>,
+        ir_module: &crate::Module,
+        ir_function: &crate::Function,
+        function: &Function,
+        block: &mut Block,
+    ) -> Result<Word, Error> {
+        let length_id =
+            self.write_runtime_array_length(array, ir_module, ir_function, function, block)?;
+
+        let limit_type = LocalType::Value {
```

Comment:
we could return this from `write_runtime_array_length` (as a part of the tuple), since it's already getting this type

Reply:
I added a helper function for producing the unsigned int type.